ServerConnection.py

```diff
@@ -0,0 +1,189 @@
+
+
+class Action(IntEnum):
+    LOGIN = 1
+    LOGOUT = 2
+    MAP = 3
+    GAME_STATE = 4
+    GAME_ACTIONS = 5
+    TURN = 6
+    CHAT = 100
+    MOVE = 101
+    SHOOT = 102
+
+
+class ServerConnection:
+    serverAddress = "wgforge-srv.wargaming.net"
+    serverPort = 443
+
+
+    def __init__(self):
+        '''
+        Opens a socket to the server.
+        '''
+        self.__Socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.__Socket.connect((self.serverAddress, self.serverPort))
+        self.__Socket.settimeout(30)
+
+
+
+    def __sendRequest(self, actionCode : int, data : jsonDict = None) -> jsonDict:
+        '''
+        Sends a request to the server and returns the response.
+
+        :param actionCode: The code of the action.
+        :param data: The request data dictionary.
+
+        :return: The response of the request.
+        '''
+        # Convert the data to a JSON string and get its length
+        if data:
+            dataJson = json.dumps(data)
+            dataLen = len(dataJson)
+        else: # No data
+            dataJson = ""
+            dataLen = 0
+
+        # Construct the message
+        msg = struct.pack("<II", actionCode, dataLen) + dataJson.encode("utf-8")
+
+        # Send the message to the server
+        self.__Socket.sendall(msg)
+
+        # Receive the response from the server
+        response = self.__Socket.recv(8)
+
+        # Unpack the response
+        resultCode, dataLen = struct.unpack("<II", response)
+
+        # Receive the response dictionary (if there's one)
+        data = ""
+        while dataLen > 0:
+            newData = self.__Socket.recv(dataLen) # receive
+            newData = newData.decode("utf-8") # decode
+            dataLen -= len(newData)
+            data += newData
+
+        if data:
+            data = json.loads(data) # turn into dictionary
+
+        # Return the response as a json dictionary
+        return {"resultCode": resultCode, "data": data}
+    
+
+    def login(self, data : jsonDict) -> jsonDict:
+        '''
+        Logs in the player to the server.\n
+
+        :param data: The login request data dictionary.
+        The server expects to receive the following required values:
+            name - player's name.
+        The following values are not required:
+            password - player's password used to verify the connection. If the player tries to connect with another password, the login will be rejected. Default: "".
+            game - game's name (use it to connect to an existing game). Default: null.
+            num_turns - number of game turns to be played. Default: null. (If num_turns is null, the default game length will be used.)
+            num_players - number of players in the game. Default: 1.
+            is_observer - defines if a player joins a server just to watch. Default: false.
+
+        :return: The response of the request.
+        '''
+        return self.__sendRequest(Action.LOGIN.value, data)
```

Comment:
sendRequest can accept enum instance and call .value internally. This will eliminate necessity to call .value all over the code.

Reply:
We didn't know that, thanks for the info 😄